### ISSUE TITLE
Apparent bug with with duplicate candidates

### ISSUE THREAD

**Issue by HEdingfield** [BrightSpots/rcv#115]
Go through `ElectionConfig.processCandidateData()` using an example with duplicate candidate names and/or candidate codes. It looks to me like `candidatePermutation` will get the value added multiple times, which distorts the probabilities during the shuffle. I think `candidatePermutation` should be populated not in the initial `for` loop that populates `candidateCodeToNameMap`, but instead via a subsequent iteration over `candidateCodeToNameMap` (since it's a HashMap which prevents duplicate values).

I also think we should take this as an opportunity to add checks for duplicate candidate names and codes in the validation method.

**Comment by HEdingfield** [comment]
Also worth noting that, even though the HashMap prevents duplicate values; if there are duplicate values, it'll subsequently overwrite whatever value is associated with that key. To me, this indicates that we want to check for duplicates and throw an error before we even run this method.

**Comment by tarheel** [comment]
Yeah, given our current design, the config should fail validation if a candidate is listed more than once. Will fix.

**Comment by HEdingfield** [comment]
@CalebKleppner please see my conversation with @tarheel below. For now I'm designing validation like he said (no duplicate names or codes allowed under any circumstances). Will keep this open until you verify that that sounds good, and we don't want to support the two cases I outlined.

> hylton [5:42 PM]
> how do we feel about different candidate names sharing the same code? should this not be allowed?
> 
> hylton [5:58 PM]
> also wondering about if there are duplicate names, but one has a code and the other doesn't; is that ok?
> 
> louis [6:16 PM]
> 1) I think we should enforce uniqueness among both names and codes.
> 2) We also should enforce that either all candidates have codes, or none do.
> 
> hylton [6:17 PM]
> uniqueness among names isn't required as long as their codes are unique though, right?
> 
> louis [6:18 PM]
> Seems hard to imagine a valid use case for duplicate names. My preference is to disallow it unless/until we hear that it's a problem.
> 
> hylton [6:18 PM]
> My understanding was that the whole point of codes was to disambiguate duplicate names, but maybe I'm wrong
> i.e. there are likely elections that occur where candidates have the same name
> 
> louis [6:19 PM]
> I think it's just to support cases in which the CVRs use codes instead of full names.
> perhaps@Caleb can weigh in here
> 
> hylton [6:20 PM]
> I'll go with what you said above for now, and will ask for clarification in the issue from him before we close it

**Comment by HEdingfield** [comment]
@CalebKleppner weighed in:

> On candidates and codes:  The main point of codes is to make the CVRs more compact (e.g., C01 instead of, "Kleppner, Caleb X (Green)").  I think it's a safe requirement that both candidates and codes be unique.  If there are two John Smiths in a race, we would call them John Smith1 and John Smith2 to keep them straight (and use codes JS1 and JS2).  Some CVRs have codes, some have candidate names.  We want to be able to handle either one.

So it sounds like this issue will be resolved once #126 gets merged in!
